cirq/experiments/cross_entropy_benchmarking.py

```diff
@@ -0,0 +1,347 @@
+import numpy as np
+from typing import List, Set, Tuple, Sequence, Dict, Any
+
+from matplotlib import pyplot as plt
+from cirq import devices, ops, circuits, sim
+
+
+class CrossEntropyResult:
+    """Results from a cross-entropy benchmarking (XEB) experiment."""
+
+    def __init__(self, num_cycle_range: Sequence[int],
+                 xeb_fidelities: Sequence[float]):
+        """
+        Args:
+            num_cycle_range: The different numbers of cycles (circuit depth) at
+                which the measurements were taken.
+            xeb_fidelities: The average XEB fidelity at each number of cycles.
+        """
+        self._num_cycles = num_cycle_range
+        self._xeb_fidelities = xeb_fidelities
+
+    @property
+    def data(self) -> Sequence[Tuple[int, float]]:
+        """Returns a sequence of tuple pairs with the first item being a
+        number of cycles and the second item being the corresponding average
+        XEB fidelity.
+        """
+        return [(num, prob) for num, prob in zip(self._num_cycles,
+                                                 self._xeb_fidelities)]
+
+    def plot(self, **plot_kwargs: Any) -> None:
+        """Plots the average XEB fidelity vs the number of cycles.
+
+        Args:
+            **plot_kwargs: Arguments to be passed to matplotlib.pyplot.plot.
+        """
+        fig = plt.figure()
+        ax = plt.gca()
+        ax.set_ylim([0, 1.1])
+
+        plt.plot(self._num_cycles, self._xeb_fidelities, 'ro-',
+                 figure=fig, **plot_kwargs)
+        plt.xlabel('Number of Cycles', figure=fig)
+        plt.ylabel('XEB Fidelity', figure=fig)
+        fig.show()
+
+
+def cross_entropy_benchmarking(
+        sampler: sim.Sampler,
+        qubits: Sequence[devices.GridQubit],
+        two_qubit_gate: ops.gate_features.TwoQubitGate,
+        *,
+        num_circuits: int = 20,
+        repetitions: int = 1000,
+        num_cycle_range: Sequence[int] = range(2, 103, 10),
+        interaction_sequence: List[Set[Tuple[devices.GridQubit,
+                                             devices.GridQubit]]] = None,
+        use_tetrahedral_group: bool = False) -> CrossEntropyResult:
+    r"""Cross-entropy benchmarking (XEB) of multiple qubits.
+
+    A total of M random circuits are generated, each of which is made of N
+    cycles where N = max(num_cycle_range). Every cycle contains randomly
+    generated single-qubit gates applied to each qubit, followed by two-qubit
+    gate(s) applied to one or more pairs of qubits. For each given number of
+    cycles n in num_cycle_range, the experiment performs the following:
+
+    1) Measure the bit-string probabilities P^(meas, m)_|...00>,
+    P^(meas, m)_|...01>, P^(meas, m)_|...10>, P^(meas, m)_|...11>... at the
+    end of circuit_mn for all m, where circuit_mn is the first n cycles of
```

Comment:
We don't actually need it for all `m`. That only makes sense for small circuits where you can get enough samples to expect to estimate all the probabilities. In practice you may not even see some of the samples.

In fact, in some XEB-like tests (namely the HOG test; Heavy Output Generation), duplicate results are discarded to avoid the possibility of gaming the test. You don't want a broken machine that always returns 000000...0 to score highly because 000...0 happened to be above average.

Reply:
Great point! I have modified the doc-string to indicate that bit-strings not seen in experiment do not factor into the estimation for `s_meas`. I think no modification of the code is necessary (correct me if I'm wrong here), since p_meas for those bit-strings are 0 (therefore gives 0 in the summation anyways).